source/agora/consensus/EnrollmentManager.d

```diff
@@ -212,7 +218,13 @@ public class EnrollmentManager
     public bool updateEnrolledHeight (const ref Hash enroll_hash,
         const size_t block_height) @safe
     {
-        if (!this.validator_set.updateEnrolledHeight(enroll_hash, block_height))
+        Enrollment enroll;
+        if (!this.enroll_pool.getEnrollment(enroll_hash, enroll))
+            return false;
+
+        this.enroll_pool.remove(enroll_hash);
```

Comment:
There is a problem here:

- We take one enrollment from the pool
- We remove it from the pool
- We try to add it to the validator set

If the last step fails the enrollment no longer exists anywhere because it was removed from the pool. We should only call `this.enroll_pool.remove(enroll_hash);` if updating the enrolled height (adding it to the validator set) was successful.

Reply:
Thanks for your comment. In the case, the function returns `false` and the ledger will terminate a node with the `assert(0)' call. The validator set can be restored with the Ledger when a node starts.

Reply:
Then the assert should be moved inside of here. Otherwise this is what's called "leaky abstractions".

Reply:
It could be better for the ledger to have the role of determining whether a node terminates or not. But, I'll think about what you suggested.

Reply:
If you want the Ledger to have the option to recover, then you need to make sure the EnrollmentManager is in a valid state.

That means to not remove the enrollment from the pool unless it was successfully added to the validator set.

Reply:
The information for validators is stored in a block and the ledger restores validators from the block. At the time when the `updateEnrolledHeight` is calling, the validators' information has already been stored in a block.

Reply:
> It could be better for the ledger to have the role of determining whether a node terminates or not.

Ideally, it would be better if the node would not terminate ;)

> At the time when the `updateEnrolledHeight` is being called, the validators' information has already been stored in a block.

Right, but we don't read the block here, we read the `enroll_pool`. And if the `enroll_pool` does not contain the enrollment, we bail out. No checking the block, no requesting other nodes.

Also, I didn't see where it was called from ?